File: src/1235-recommend/rk1.ts
import * as alcoholMap from '../../docs/alcohol-map.json';
import * as readlineSync from 'readline-sync';
import * as fs from 'fs';
import {simpleDeepCopy} from "./algorithms/copy";
import {readNaturalOrZeroNumberOf, readProducts} from "./io/readProduct";
import {recommend} from "./recommender/recommend";
import {ApproxSearchEngine} from './search/approx';
import {LeafData, ExtendedLeafData} from './types/data';
import {TreeNode} from './types/node';
import {getLeavesData, normalizeLeavesData} from './algorithms/leaves';
import {SearchConfig} from './types/search';
import {SearchEngine} from './search/search';


class RK1 {
	root: TreeNode;
	leavesData: LeafData[];
	extendedLeavesData: ExtendedLeafData[];
	workingLeavesData: ExtendedLeafData[];

	constructor(root: TreeNode) {
		this.root = root;
		this.leavesData = getLeavesData(root);
		this.extendedLeavesData = normalizeLeavesData(this.leavesData);
		this.workingLeavesData = simpleDeepCopy(this.extendedLeavesData);
	}

	run() {
		while (true) {
			console.log('0. Reset All');
			console.log('1. Recommend (LAB 3)');
			console.log('2. Search (LAB 5)');
			console.log('3. Quit');
			const n = readNaturalOrZeroNumberOf('menu', 3);
			switch (n) {
				case 0:
					this.reset();
					break;
				case 1:
					this.recommend();
					break;
				case 2:
					this.search();
					break;
				case 3:
					console.log('Good bye!');
					process.exit();
			}
		}
	}

	reset() {
		this.workingLeavesData = simpleDeepCopy(this.extendedLeavesData);
		console.clear();
	}

	recommend() {
		const products: string[] = this.workingLeavesData.map((leafData: ExtendedLeafData): string => leafData.source.name);
		console.log('List of products: ', products);

		const likes = readProducts('likes', products);
		const productsMinusLikes = products.filter((product: string): boolean => !likes.includes(product));

		const dislikes = readProducts('dislikes', productsMinusLikes);

		const recommendations = recommend(this.root, this.workingLeavesData, likes, dislikes);
		const n = readNaturalOrZeroNumberOf('out recommendations', recommendations.length);
		this.workingLeavesData = recommendations.slice(0, n);
		console.log(this.workingLeavesData);
	}

	search() {
		readlineSync.question('Change search-config.json and press any key to search');
		const searchConfigs: SearchConfig[] = JSON.parse(fs.readFileSync('./docs/search-config.json').toString());
		console.log(searchConfigs);
		const searchEngine: SearchEngine = new SearchEngine(searchConfigs);

		const searchResults = searchEngine.findAll(this.workingLeavesData);
		if (!searchResults.length) {
			const approxSearchEngine = new ApproxSearchEngine(searchConfigs, this.leavesData);
			console.log('========== No exact match found, however you may like :');
			const n = readNaturalOrZeroNumberOf('out recommendations', this.workingLeavesData.length);
			this.workingLeavesData = approxSearchEngine.findAll(this.root, this.workingLeavesData).slice(0, n);
			console.log(this.workingLeavesData);
			return;
		}
		console.log('========== Found:');
		console.log(searchResults);
		this.workingLeavesData = searchResults;
	}
}


const main = () => {
	const root: TreeNode = alcoholMap;

	const rk1 = new RK1(root);
	rk1.run();
};

main();

File: src/678-dialog/rk1.py
from copy import deepcopy
from sys import exit
from typing import List

from algorithms.leaves import get_leaves_data, normalize_leaves_data
from io_util.read_json import read_map, read_config
from io_util.read_product import read_natural_or_zero_number_of, read_products
from recommender.recommend import recommend
from search.approx import ApproxSearchEngine
from search.search import SearchEngine
from ttypes.data import LeafData, ExtendedLeafData
from ttypes.node import TreeNode
from ttypes.search import SearchConfig


class RK1:
    root: TreeNode
    leaves_data: List[LeafData]
    extended_leaves_data: List[ExtendedLeafData]
    working_leaves_data: List[ExtendedLeafData]

    def __init__(self, root: TreeNode):
        self.root = root
        self.leaves_data = get_leaves_data(root)
        self.extended_leaves_data = normalize_leaves_data(self.leaves_data)
        self.working_leaves_data = deepcopy(self.extended_leaves_data)

    def run(self) -> None:
        while True:
            print('0. Reset All')
            print('1. Recommend (LAB 3)')
            print('2. Search (LAB 5)')
            print('3. Quit')
            n = read_natural_or_zero_number_of('menu', 3)
            match n:
                case 0:
                    self.reset()
                case 1:
                    self.recommend()
                case 2:
                    self.search()
                case 3:
                    print('Good bye!')
                    exit()

    def reset(self):
        self.working_leaves_data = deepcopy(self.extended_leaves_data)
        print(end='\033[H\033[J')

    def recommend(self):
        products: List[str] = [leaf_data.source['name'] for leaf_data in self.working_leaves_data]
        print('List of products: ', products)

        likes = read_products('likes', products)
        products_minus_likes = [product for product in products if product not in likes]

        dislikes = read_products('dislikes', products_minus_likes)

        recommendations = recommend(self.root, self.working_leaves_data, likes, dislikes)
        n = read_natural_or_zero_number_of('out recommendations', len(recommendations))
        self.working_leaves_data = recommendations[:n]
        print(self.working_leaves_data)

    def search(self):
        _ = input('Change search-config.json and press any key to search')
        search_configs: List[SearchConfig] = read_config()
        print(search_configs)
        search_engine: SearchEngine = SearchEngine(search_configs)

        search_results = search_engine.find_all(self.working_leaves_data)
        if not search_results:
            approx_search_engine = ApproxSearchEngine(search_configs, self.leaves_data)
            print('========== No exact match found, however you may like :')
            n = read_natural_or_zero_number_of('out recommendations', len(self.working_leaves_data))
            self.working_leaves_data = approx_search_engine.find_all(self.root, self.working_leaves_data)[:n]
            print(self.working_leaves_data)
            return
        print('========== Found:')
        print(search_results)
        self.working_leaves_data = search_results


def main():
    root: TreeNode = read_map()

    rk1 = RK1(root)
    rk1.run()


if __name__ == '__main__':
    main()
